pkg/shifter/shifter.go

```diff
@@ -0,0 +1,255 @@
+		case <-ctx.Done():
+			ss.logger.Info("Shutting down Shifter service")
+			wg.Done()
+			return
+		default:
+			// Check if current time is in bettween whiteListIntervals. If yes, run ss.shift().
+			t := time.Now().UTC()
+			hour := fmt.Sprintf("%02d", t.Hour())
+			mins := fmt.Sprintf("%02d", t.Minute())
+			fmtTime := fmt.Sprintf("%s:%s", hour, mins)
+			now, err := time.Parse(timeLayout, fmtTime)
+
+			if err != nil {
+				panic(err)
+			}
+
+			for _, interval := range ss.whiteListIntervals {
+				if timeWithinWLIntervalCheck(interval.start, interval.end, now) {
+					ss.shift()
+				}
+			}
+			time.Sleep(time.Millisecond * time.Duration(ss.cp.GetInt(config.ShifterPollIntervalMs)))
+		}
+	}
+}
+
+//getNodePoolMap finds out fallback on-demand nodePools and their respective preemptible node-pools.
+func (ss *ShifterService) getNodePoolMap() error {
+
+	nps, err := ss.gcloudClient.ListNodePools()
+
+	ss.logger.Info(fmt.Sprintf("Nodepools in the cluster %v", nps))
+
+	if err != nil {
+		return err
+	}
+
+	var preemptibleNodePools, onDemandNodePools []container.NodePool
+
+	for _, np := range nps {
+		if np.Config.Preemptible {
+			preemptibleNodePools = append(preemptibleNodePools, *np)
+		} else {
+			onDemandNodePools = append(onDemandNodePools, *np)
+		}
+	}
+
+	// Loop over preemptible nodepools and find their corresponding fallback on-demand node-pool.
+	ss.nodePoolMap = make(map[string]npShiftConf)
+	for _, pnp := range preemptibleNodePools {
+		for _, onp := range onDemandNodePools {
+			if reflect.DeepEqual(pnp.Config.Labels, onp.Config.Labels) && pnp.Config.MachineType == onp.Config.MachineType {
+
+				ss.nodePoolMap[onp.Name] = npShiftConf{
+					destNodePool:       pnp.Name,
+					sourceMinNodeCount: onp.Autoscaling.MinNodeCount,
+				}
+			}
+		}
+	}
+	return nil
+}
+
+//getNodePoolSize gets the node-pool size by checking maximum number of nodes in the available zones.
+//This works for single and multi zone clusters.
+func (ss ShifterService) getNodePoolSize(selector string) (int64, error) {
+	nodes, err := ss.kubeClient.GetNodes(selector)
+
+	if err != nil {
+		return 0, err
+	}
+
+	nodeZoneWise := make(map[string]int64)
+
+	for _, node := range nodes.Items {
+		zone := node.ObjectMeta.Labels["failure-domain.beta.kubernetes.io/zone"]
+		if _, ok := nodeZoneWise[zone]; ok {
+			nodeZoneWise[zone]++
+		} else {
+			nodeZoneWise[zone] = 1
+		}
+	}
+
+	var max int64
+
+	for _, count := range nodeZoneWise {
+		if count > max {
+			max = count
+		}
+	}
+
+	return max, err
+}
+
+func (ss ShifterService) makeNodeUnschedulable(nodes []v1.Node) error {
+
+	for _, node := range nodes {
+		ss.logger.Info(fmt.Sprintf("Cordoning node %v", node.Name))
+		recentNodeObject, err := ss.kubeClient.GetNode(node.Name)
+		if err != nil {
+			return err
+		}
+		recentNodeObject.Spec.Unschedulable = true
+		err = ss.kubeClient.UpdateNode(recentNodeObject)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (ss ShifterService) shift() {
```

Comment:
We should have a test for this function as the main logic belongs here. Assertions on the calls we are supposed to make as part if this will be enough I guess. Killer and Gcloud clients can be mocked.

Reply:
Added